Docs for Cubical indexed types

plt-amy opened the pull request:
PR just because I've had bad times with reST before, and it would be embarrassing to mess that up on master. (Closes #6227)

plt-amy committed:
docs: Cubical indexed types

andreasabel commented:
Rendering here: https://agda--6232.org.readthedocs.build/en/6232/language/cubical.html

andreasabel commented on doc/user-manual/language/cubical.lagda.rst:
```diff
@@ -1,7 +1,7 @@
   ::
 
-  {-# OPTIONS --cubical #-}
+  {-# OPTIONS --cubical -vtc.lhs.unify.inv:20 -WNoEquivWhenSplitting #-}
```
Do these extra options have any effect on the user-manual or its build?

andreasabel replied:
I removed them in my fixup, but feel free to restore them if that wasn't good.

plt-amy replied:
Ah, thanks; they were just to help me figure out a minimal reproducer for the Setω transport case. (I think I'm going to write a pre-commit hook that looks through the diff and asks if I really meant to commit the verbosity flags....)

plt-amy committed:
fixup: link formatting

andreasabel reviewed (commented):
I committed some fixes directly.

andreasabel commented:
Er, I did the edit directly on github, and I think it screwed up when your commit came inbetween.

Let me repair this...

andreasabel committed:
fixup: formatting

andreasabel commented:
Ok, fixed my commit.

andreasabel reviewed (approved):
From my side, this looks good, modulo I do not really understand Cubical enough to fully follow.  Maybe @mortberg wants to review...

andreasabel commented:
Looks like we rather need option `-WnoNoEquivWhenSplitting` (mmh, doubly-negated warning):
```
UserManual
    language-cubical:                              FAIL (0.89s)
                                                   Result did not match golden value.
                                                   Diff between actual and golden value:
...
+Could not generate equivalence when splitting on indexed family,
+the function will not compute on transports by a path.
+  Reason: UnsupportedYet Injectivity
+                           position:    0
+                           type:        Nat
+                           datatype:    Nat
+                           parameters: 
+                           indices:    
+                           constructor: suc
+when checking the definition of sucInjEq
```

andreasabel commented:
Let me fix the conflict I caused.

andreasabel committed:
Merge branch 'master' into aliao/transpix-docs, link to #3733

Conflicts:
	CHANGELOG.md

andreasabel commented:
I merged in `master`, we need to __SQUASH__ when merging this PR.

andreasabel committed:
Re #6232: add -WnoNoEquivWhenSplitting

andreasabel commented on doc/user-manual/language/cubical.lagda.rst:
```diff
@@ -726,8 +729,143 @@ is defined as:
+    _ = refl
+
+The computational behaviour of matching on indexed inductives in very
+generic is suboptimal when compared to the behaviour of similar
```
I don't see the transition of thought from "suboptimal" (negative) to the example that follows, where "we can investigate" (positive).  There is a gap in the argument, imo.

Can you reformulate to clarify?

andreasabel commented on doc/user-manual/language/cubical.lagda.rst:
```diff
@@ -726,8 +729,143 @@ is defined as:
+
+In more concrete situations, such as when the indices are constructors
+of some other inductive type, pattern-matching definitions will not
+compute when applied to transports.
```
Would this warrant an example?

andreasabel replied:
Ah I see this comes later, but is interrupted by the `NoEquivWhenSplitting` discourse.

Please add the usual forward referencing.

andreasabel commented on doc/user-manual/language/cubical.lagda.rst:
```diff
@@ -726,8 +729,143 @@ is defined as:
+unification produces something that can not be extended to cover
+transports, and the cases in which it can. The following pair of
+definitions relies on injectivity for data constructors, and so will not
+compute on non-``reflEq`` paths:
```
Where will this hit me?  E.g. what lemma on `sucInjEq` will then fail to go through?

andreasabel commented on doc/user-manual/language/cubical.lagda.rst:
```diff
@@ -726,8 +729,143 @@ is defined as:
+In some cases, it may be possible to rephrase the proof in ways that
+avoid pattern matching, and so, compute. As a simple example, since
+``apEq`` below will always compute, we can use the fact that ``suc`` has
+a partially-defined inverse to conclude it is injective:
```
I remember this trick from my apprentice years in type theory.

This seems to document a workaround.  Ok.  Might a good advice to users...

andreasabel reviewed (changes requested):
I sat down again in a quieter moment and can say that I mostly follow this exposition.
I added some comments, some with a request to clarify the formulation.
Thanks for the docs!

plt-amy committed:
doc: demonstrate failure of computation on transports

the "--Continuing from above..." line is necessary for Sphinx to
preserve the indentation of the _ : hd ≡ true lines, at least on my
machine (using make user-manual-html).

andreasabel reviewed (approved):
Great job, thanks!
